flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/ColumnarArray.java

```diff
@@ -0,0 +1,332 @@
+	}
+
+	@Override
+	public void setNullBoolean(int pos) {
+		throw new UnsupportedOperationException("Not support the operation!");
+	}
+
+	@Override
+	public void setNullByte(int pos) {
+		throw new UnsupportedOperationException("Not support the operation!");
+	}
+
+	@Override
+	public void setNullShort(int pos) {
+		throw new UnsupportedOperationException("Not support the operation!");
+	}
+
+	@Override
+	public void setNullFloat(int pos) {
+		throw new UnsupportedOperationException("Not support the operation!");
+	}
+
+	@Override
+	public void setNullDouble(int pos) {
+		throw new UnsupportedOperationException("Not support the operation!");
+	}
+
+	@Override
+	public boolean[] toBooleanArray() {
+		boolean[] res = new boolean[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getBoolean(i);
+		}
+		return res;
+	}
+
+	@Override
+	public byte[] toByteArray() {
+		byte[] res = new byte[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getByte(i);
+		}
+		return res;
+	}
+
+	@Override
+	public short[] toShortArray() {
+		short[] res = new short[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getShort(i);
+		}
+		return res;
+	}
+
+	@Override
+	public int[] toIntArray() {
+		int[] res = new int[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getInt(i);
+		}
+		return res;
+	}
+
+	@Override
+	public long[] toLongArray() {
+		long[] res = new long[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getLong(i);
+		}
+		return res;
+	}
+
+	@Override
+	public float[] toFloatArray() {
+		float[] res = new float[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getFloat(i);
+		}
+		return res;
+	}
+
+	@Override
+	public double[] toDoubleArray() {
+		double[] res = new double[length];
+		for (int i = 0; i < length; i++) {
+			res[i] = getDouble(i);
+		}
+		return res;
+	}
+
+	public BaseArray copy() {
+		LogicalType elementLogicalType = elementType.getLogicalType();
+		if (elementLogicalType instanceof BooleanType) {
+			return BinaryArray.fromPrimitiveArray(toBooleanArray());
+		} else if (elementLogicalType instanceof TinyIntType) {
+			return BinaryArray.fromPrimitiveArray(toByteArray());
+		} else if (elementLogicalType instanceof SmallIntType) {
+			return BinaryArray.fromPrimitiveArray(toShortArray());
+		} else if (elementLogicalType instanceof IntType || elementLogicalType instanceof DateType || elementLogicalType instanceof TimeType) {
+			return BinaryArray.fromPrimitiveArray(toIntArray());
+		} else if (elementLogicalType instanceof BigIntType) {
+			return BinaryArray.fromPrimitiveArray(toLongArray());
+		} else if (elementLogicalType instanceof FloatType) {
+			return BinaryArray.fromPrimitiveArray(toFloatArray());
+		} else if (elementLogicalType instanceof DoubleType) {
+			return BinaryArray.fromPrimitiveArray(toDoubleArray());
+		} else {
+			DataFormatConverters.ObjectArrayConverter converter = new DataFormatConverters.ObjectArrayConverter(elementType);
```

Comment:
I can not get this. What we want is copy instead of conversion between external and internal.